sites/all/modules/ding_unilogin/ding_unilogin.module

```diff
@@ -0,0 +1,342 @@
+function _ding_unilogin_handle_login() {
+  if (isset($_GET['user']) &&
+      isset($_GET['timestamp']) &&
+      isset($_GET['auth'])) {
+    $error = FALSE;
+    $success = FALSE;
+    // Validate ticket.
+    $hash = md5($_GET['timestamp'] .
+                variable_get('ding_unilogin_secret', '') .
+                $_GET['user']);
+    if ($hash === $_GET['auth']) {
+      $time = DateTime::createFromFormat('YmdHis', $_GET['timestamp'],
+                                         new DateTimeZone('UTC'));
+
+      if ($time) {
+        // Check that the given time stamp is in the past (with 5 seconds
+        // grace to avoid NTP race conditions), but not older than
+        // one minute.
+        $now = (new DateTime())->getTimestamp();
+        if ($time->getTimestamp() < ($now + 5) &&
+            ($now - $time->getTimestamp()) < 60
+        ) {
+          $success = _ding_unilogin_login($_GET['user']);
+        }
+        else {
+          $error = 'Timestamp %time too old for %user';
+        }
+
+      }
+      else {
+        $error = 'Could not parse timestamp %time for %user';
+      }
+    }
+    else {
+      $error = "Auth hash din't validate for %user";
+    }
+
+    if ($error) {
+      watchdog('ding_unilogin', $error,
+               array('%time' => $_GET['timestamp'], '%user' => $_GET['user']),
+               WATCHDOG_WARNING);
+      drupal_set_message(t('Could not understand the reply from UNI•Login, so cannot log you in.'), 'error');
+    }
+    // Redirect to current_path() to clear most GET parameters.
+    $query = array();
+    if (isset($_GET['ding-unilogin-trigger']) && $success) {
+      // This should propagate if we logged in a user, else it doesn't matter
+      // (doesn't make sense to re-display the reservation popup, as it'll
+      // just trigger the login form again).
+      $query['ding-unilogin-trigger'] = $_GET['ding-unilogin-trigger'];
+    }
+    drupal_goto(current_path(), array('query' => $query));
+  }
+}
+
+/**
+ * URL to redirect to for UNI•Login.
+ */
+function ding_unilogin_url($path = NULL) {
+  $unilogin_id = variable_get('ding_unilogin_id', '');
+
+  // If no path given default to redirecting back to current path.
+  if (!$path && request_uri() != '/') {
+    $path = url(request_uri(), array('absolute' => TRUE));
+  }
+
+  if (!$path) {
+    return sprintf('https://sso.emu.dk/unilogin/login.cgi?id=%s', $unilogin_id);
+  }
+
+  $secret = variable_get('ding_unilogin_secret', '');
+  return sprintf('https://sso.emu.dk/unilogin/login.cgi?id=%s&path=%s&auth=%s',
+                 $unilogin_id,
+                 urlencode(base64_encode($path)),
+                 md5($path . $secret)
+  );
+}
+
+/**
+ * Check whether a user is logged in via UNI•Login.
+ */
+function ding_unilogin_authenticated($account = NULL) {
+  global $user;
+  if (!$account) {
+    $account = $user;
+  }
+
+  if (isset($account->data['unilogin']) && $account->data['unilogin']) {
+    return TRUE;
+  }
+  return FALSE;
```

Comment:
This could be simplified to `return isset($account->data['unilogin']) && $account->data['unilogin'])`.